scvi/dataset/anndata.py

```diff
@@ -79,13 +80,17 @@ def extract_data_from_anndata(self, ad: anndata.AnnData):
             ad.obs
         )  # provide access to observation annotations from the underlying AnnData object.
 
+        # treat all possible cases according to anndata doc
         if isinstance(ad.X, np.ndarray):
-            data = ad.X.copy()  # Dense
-        else:
-            data = ad.X.toarray()  # Sparse
-
-        select = data.sum(axis=1) > 0  # Take out cells that doesn't express any gene
-        data = data[select, :]
+            data = ad.X.copy()
+        if isinstance(ad.X, pd.DataFrame):
+            data = ad.X.values
+        if isinstance(ad.X, csr_matrix):
+            # keep sparsity above 100 Mb
+            if ad.X.data.nbytes < 1e8:
+                data = ad.X.toarray()
```

Comment:
How about deciding whether to keep sparsity based on the size the dense matrix would be? That seems most relevant. I'd say, switch to dense format always if the dense matrix would be less than 1 GB.

Reply:
haha obviously, I was tired..